### ISSUE TITLE
Review projection APIs

### ISSUE THREAD

**Issue by bounoable** [modernice/goes#34]
This issue will determine if the different projection APIs are even necessary to solve common problems or if there are better solutions to these problems that don't require these APIs.

## HistoryDependent

The `HistoryDependent` API is used by projection jobs to determine which events a projection needs to properly update itself. If a projection implements HistoryDependent, it can hint to a projection job that it requires the full history of the events that are configured in the job, instead of just the (published) events that triggered the job.

## ProgressAware

The `ProgressAware` API is usually used by persistent projections that are stored in a database. When a projection job is triggered, the application fetches the current projection state from the database, applies the events on it and saves it back to the database. A ProgressAware projection keeps track of the last applied event in terms of the event time, to ensure that no event is applied twice to the projection.

**Comment by bounoable** [comment]
## ProgressAware – Scenario: *X*

Given the same ecommerce app as before but with an `order` service. A read-model for is projected for each order.

### Code Example

```go
package order

const (
  Placed = "order.placed"
  Canceled = "order.canceled"
)

type PlacedData struct {
   Items []Item
}

type Order struct {
  *aggregate.Base
  OrderDTO
}

type OrderDTO struct {
  ID uuid.UUID
  Items []Item
}

type Item struct {
  ProductID uuid.UUID
  Quantity uint
  UnitPrice uint64
}

// Domain logic
func (o *Order) Place(items []Item) error { ... }
func (o *Order) Cancel() error { ... }

type ReadModelRepository = model.Repository[*OrderRM, uuid.UUID]

// OrderRM is the read-model of an order.
type OrderRM struct {
  OrderDTO
}

type Projector struct {
  orders ReadModelRepository
  schedule *schedule.Continuous
}

func NewProjector(orders ReadModelRepository, bus event.Bus, store event.Store) *Projector {
  return &Projector{
    orders: orders,
    schedule: schedule.Continuously(bus, store, []string{Placed, Canceled}),
  }
}

func (proj *OrderProjector) Run(ctx context.Context) (<-chan error, error) {
  return proj.schedule.Subscribe(ctx, proj.applyJob)
}

func (proj *OrderProjector) applyJob(ctx projection.Job) error {
  orders, errs, err := proj.extractOrdersFromJob(ctx)
  if err != nil {
    return fmt.Errorf("extract orders from job: %w")
  }

  for _, orderID := range orders {
    if err := proj.orders.Use(ctx, orderID, func(o *OrderRM) {
      return ctx.Apply(ctx, o)
    }); err != nil {
      log.Printf("failed to project order: %v [order=%s]", err, orderID)
    }
  }
}
```

### Problem

Currently, the ProgressAware API is kind of a no-op because projection jobs disable progress checks when applying event onto a projection. This is due to

**Comment by bounoable** [comment]
Closed by #36
